lib/hammer_cli_foreman/auth.rb

```diff
@@ -1,34 +1,70 @@
-module HammerCLIForeman
+require 'hammer_cli_foreman/openid_connect.rb'
+require 'hammer_cli_foreman/authenticate/login.rb'
 
+module HammerCLIForeman
   class Auth < HammerCLI::AbstractCommand
-
     class LoginCommand < HammerCLI::AbstractCommand
-      command_name "login"
+      command_name 'login'
       desc _("Set credentials")
 
-      option ["-u", "--username"], "USERNAME", _("Username to access the remote system")
-      option ["-p", "--password"], "PASSWORD", _("Password to access the remote system")
+      class Basic < HammerCLI::AbstractCommand
+        extend HammerCLIForeman::Authenticate::Login
 
-      def execute
-        if !(HammerCLIForeman.foreman_api_connection.authenticator.is_a?(HammerCLIForeman::Api::SessionAuthenticatorWrapper))
-          print_message(_("Can't perform login. Make sure sessions are enabled in hammer configuration file."))
-          return HammerCLI::EX_USAGE
+        command_name('basic')
+        desc('provide username and password')
+
+        option ["-u", "--username"], "USERNAME", _("Username to access the remote system")
+        option ["-p", "--password"], "PASSWORD", _("Password to access the remote system")
+
+        def execute
+          Basic.execute_with_params(
+            AUTH_TYPES[:basic_auth],
+            option_username || HammerCLI::Settings.get('_params', 'username'),
+            option_password || HammerCLI::Settings.get('_params', 'password')
+          )
         end
+      end
 
-        # Make sure we reflect also credentials set for the main hammer command
-        # ( hammer -u test auth login )
-        HammerCLIForeman.foreman_api_connection.authenticator.set_credentials(
-          option_username || HammerCLI::Settings.get('_params', 'username'),
-          option_password || HammerCLI::Settings.get('_params', 'password')
-        )
-        HammerCLIForeman.foreman_api_connection.authenticator.force_user_change
-        HammerCLIForeman.foreman_api_connection.login
+      class Oauth < HammerCLI::AbstractCommand
+        extend HammerCLIForeman::Authenticate::Login
 
-        print_message(_("Successfully logged in as '%s'.") % HammerCLIForeman.foreman_api_connection.authenticator.user)
-        HammerCLI::EX_OK
+        command_name('oauth')
+        desc('supports for both with/without 2fa')
+
+        option ["-u", "--username"], "USERNAME", _("Username to access the remote system")
+        option ["-p", "--password"], "PASSWORD", _("Password to access the remote system")
+        option ["-ote", "--openidc-token-endpoint"], "URL", _("Openidc Provider URL which issues access token")
```

Comment:
This will not work, since clamp uses `-` for single character options.

I would suggest:
 - `-t` for `--openidc-token-endpoint`
 - `-a` for `--openidc-authorizarion-endpoint`
 - `-f` for `--two-factor`
 - `-r` for `--redirect-uri`